Add new integration for Leviosa zone

greg-ha-1990 opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add new integration for Leviosa zone


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/29230

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

greg-ha-1990 committed:
Add new integration for Leviosa zone

greg-ha-1990 committed:
[leviosa_shades] fix tests error

greg-ha-1990 committed:
Code optimization

greg-ha-1990 committed:
Update tests for leviosa_shades

greg-ha-1990 committed:
Update code tests to be fully covered

greg-ha-1990 committed:
Update hostname for leviosa_shades

greg-ha-1990 committed:
Update test ip for leviosa_shades

iMicknl commented on homeassistant/components/leviosa_shades/__init__.py:
```diff
@@ -0,0 +1,29 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
```
```suggestion
    await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
```

greg-ha-1990 replied:
Suggestion accepted, thanks.

iMicknl commented on homeassistant/components/leviosa_shades/config_flow.py:
```diff
@@ -0,0 +1,156 @@
+class LeviosaConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Manages the interaction with user when a Leviosa Zone needs to be setup."""
+
+    # The schema version below will be used by Home Assistant to determine
+    # if a call to the migrate method is needed; this is not implemented
+    # as of March 2021
```
```suggestion
```

greg-ha-1990 replied:
Suggestion accepted, thanks.

iMicknl commented on homeassistant/components/leviosa_shades/config_flow.py:
```diff
@@ -0,0 +1,156 @@
+    # if a call to the migrate method is needed; this is not implemented
+    # as of March 2021
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
```
```suggestion
```

This is now part of the manifest

greg-ha-1990 replied:
Suggestion accepted, thanks.

iMicknl commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,180 @@
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the Leviosa shade groups."""
+    _LOGGER.debug(
+        "Setting up %s[%s]: %s",
+        entry.domain,
+        entry.title,
+        entry.entry_id,
+    )
+    hub_name = entry.title
+    hub_mac = entry.data["device_mac"]
+    hub_ip = entry.data["host"]
+    blind_groups = entry.data[BLIND_GROUPS]
+    _LOGGER.debug("Groups to create: %s", blind_groups)
+    hub = tZoneHub(
+        hub_ip=hub_ip, hub_name=hub_name, websession=async_get_clientsession(hass)
+    )
+    await hub.getHubInfo()  # Check all is good
+    _LOGGER.debug("Hub object created, FW: %s", hub.fwVer)
+    entities = []
+    for blind_group in blind_groups:
+        _LOGGER.debug("Adding blind_group: %s", blind_group)
+        new_group_obj = hub.AddGroup(blind_group)
+        entities.append(
+            LeviosaBlindGroup(
+                hass, hub_mac + "-" + str(new_group_obj.number), new_group_obj
+            )
+        )
+    async_add_entities(entities)
```
Please move this to your __init__.py. The logic for connecting should not be part of your cover.py.

See https://github.com/home-assistant/core/blob/dev/homeassistant/components/overkiz/__init__.py (and many other integration) for an example.

greg-ha-1990 replied:
Thank you for your suggestion.
My understanding is that since a leviosa shades can spawn multiple entities, this curtain controller is essentially a wifi to IR device, but it can control up to six groups of covers, so it will derive up to seven entities, such as `group1` - `group6` and `group_ALL`, so we need to call the `async_add_entities` method inside the `async_setup_entry` method, but obviously the `async_setup_entry` method in `__init__.py` doesn't have an `AddEntitiesCallback` parameter, so it's not possible to call `async_add_entities` inside `async_setup_entry` in __init__,py anymore.
I'm currently stuck at this point and don't know how to proceed from here, again I'll ask you for your advice, thanks!

greg-ha-1990 replied:
By the way, I looked at your sample code and I see that your scenario is logging into a cloud account and then being able to register multiple `devices`, not `entities`, And the situation I encountered is that one `device` supports multiple `entities`

greg-ha-1990 replied:
I refer to: https://github.com/home-assistant/core/blob/7d45227a5966bec0d265026dd36bd83df7e46cff/homeassistant/components/acmeda/cover.py#L22C13-L22C13

iMicknl replied:
@greg-ha-1990 in the example you link to, the hub is instantiated in `__init__.py` and retrieved via the `hass.data`. This is the same as my suggestion.

````python3
    hub: PulseHub = hass.data[DOMAIN][config_entry.entry_id]
```

Please don't resolve conversations that you didn't fix. This makes it hard for reviewers to look at the open issues.

iMicknl replied:
Yes. And in `__init__` you should check if you can connect, and otherwise catch the error / raise ConfigEntryNotReady (for retry) or another status.

Cover.py should only be responsible for setting up the entity, not all the connections around this.

greg-ha-1990 replied:
Okay, I have updated the code and the test passes, but I don’t know whether to click `Resolve conversation` at this time, so I won’t click this button yet.

iMicknl replied:
You didn't push your changes yet?

In essence, it is fine to use 'Resolve conversation' if you are sure this issue is fixed. If you aren't sure, just leave it open and a reviewer will close it when applicable. If you just close reviews you don't agree with / have different opinions on, it is hard to keep track of the discussion and open feedback.

greg-ha-1990 replied:
I just pushed the code

iMicknl commented on homeassistant/components/leviosa_shades/services.yaml:
```diff
@@ -0,0 +1,13 @@
+next_down_pos:
```
Best to remove this from your initial PR. It is not required as base functionality and can be added in a follow-up PR.

In the future, it would be good to add comments here to clarify what this does. See https://developers.home-assistant.io/docs/dev_101_services/#service-descriptions.

greg-ha-1990 replied:
There is ambiguity here, what it really means is for the motor of the cover to advance the cover to the next `position`, I will update the naming of this service to make it easier to understand and avoid ambiguity.

iMicknl replied:
Best to add this to a follow-up PR to keep the initial PR as small as possible.

greg-ha-1990 replied:
Are you referring to removing service.yaml from the current PR?

iMicknl replied:
Best is to remove all custom services. The smaller you keep the initial PR, the faster it can be merged to core. For your initial PR keep only the bare minimum required to connect to your device and one single entity. In a follow-up PR you can add additional sensors and services.

greg-ha-1990 replied:
ok, I just removed all services

iMicknl commented on homeassistant/components/leviosa_shades/config_flow.py:
```diff
@@ -0,0 +1,156 @@
+        """Perform discovery and present an input screen for each Zone discovered."""
+
+        _LOGGER.debug("Looking for Leviosa Zone HUBs")
+        self._devices = await discover_leviosa_zones()
```
What kind of discovery is this? Normally we should not use the user flow for discovery, but the existing SSDP / DHCP discovery.

greg-ha-1990 replied:
We use the SSDP protocol for device discovery, and I know HA already supports SSDP-based auto-discovery, but I still prefer to put it inside leviosa_shades instead of reusing the public SSDP discovery functionality because I want to customize the presentation of the device discovery page in a way that can help users understand the difference between different leviosa shades, and I don't want to scatter discovered leviosa shades in a massive list of chanting HA auto-discovered devices; in fact, I'll even hit the `ignore all` button，And sometimes the public device discovery function cannot complete all device discovery well.

Can I keep it in my own integration?

iMicknl replied:
No, you should use the built-in SSDP flow for this. You can set a name / description for any discovered device, thus you are able to customize the presentation of your devices.

You can still add a regular config flow where people can add the integration manually (by IP e.g.) if that is possible.

iMicknl replied:
@bdraco any advice from you here? I am not really familiar with SSDP discovery and what we support.

Custom SSDP code from underlying library:
https://github.com/altersis/aioleviosa/blob/6d09ce3f42decd9012246bad75afafd0fdb31fb0/aioleviosa/aioleviosa.py#L29-L71

emontnemery replied:
@greg-ha-1990 can you try to better explain what problem is caused if you use the built-in discovery?
Is the problem that if the user has 8 hubs, they would see something like this (although obviously for leviosa, not for shelly):
![image](https://github.com/home-assistant/core/assets/14281572/d472cca0-89e7-4578-a075-f0823b824a56)

In a previous reply, you mentioned you want to present the discovered hubs in a way which is helpful to the user, but in the code it seems like you just generate a list of IP-addresses?

greg-ha-1990 replied:
> can you try to better explain what problem is caused if you use the built-in discovery?

There is no way to automatically discover curtains on the Devices & Services homepage. This is the current problem.

> In a previous reply, you mentioned you want to present the discovered hubs in a way which is helpful to the user, but in the code it seems like you just generate a list of IP-addresses?

Of course the user experience is not good. First of all, on the Devices & Services homepage, automatically discovered devices always pop up and prompt the user to configure them. This process is completely automated in the background and discovered silently. There is no progress bar or prompt. My curtains , has not been found, I don't know why. This is the so-called poor user experience, with no errors and no prompts. I'm tired of automatic discovery, so I added a "search process" pop-up window to my integration, with a 20-second timeout. If no search is found, an error will be prompted. This is a user-friendly design. As for what you said Only the IP is displayed, because in this scenario, the only thing that needs to be concerned about is the IP, the so-called MAC, and the serial number. Not to mention that only the IP and URN can be found through SSDP. Even if the manufacturer encapsulates the MAC in the URN, then what? What's the point? The IP of the curtain has changed, but the MAC remains unchanged, and it still cannot be controlled. We must understand that we are working in a LAN scenario, and we must accurately know the IP address of the curtain in order to control it.

greg-ha-1990 replied:
> We want to avoid running multiple SSDP listeners as we already have a performance issue with processing traffic volume on some systems.
> 
> Another option is to register for a callback. Example:
> 
> https://github.com/home-assistant/core/blob/017b2623e8ff741276902578f32aa896c89eed44/homeassistant/components/sonos/__init__.py#L583

Thank you for your reply. The main problem I have is that I don't find these curtains on the homepage. They don't appear like they appear in your screenshots? I would like to add a progress bar with a 20 second timeout to prompt the user for an error. Additionally, the SSDP discovery process will only be enabled during configuration, so the efficiency issue may not be a bottleneck. In order to add a 20 second timeout error message, can I use SSDP discovery in my integration?

iMicknl commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,180 @@
+
+        return device_info
+
+    @property
+    def is_opening(self):
+        """Is the blind group opening?."""
+
+        return False
+
+    @property
+    def is_closing(self):
+        """Is the blind closing?."""
+
+        return False
```
```suggestion
```

greg-ha-1990 replied:
Suggestion accepted, thanks.

iMicknl commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,180 @@
+        """Initialize the shade group."""
+        self._blind_group_id = blind_group_id
+        self._blind_group_obj = blind_group_obj
+        self._hass = hass
```
```suggestion
        self._hass = hass
        
        self._attr_name = self._blind_group_obj.name
        self._attr_unique_id = self._blind_group_id
        self._attr_device_class = cover.DEVICE_CLASS_SHADE
        self._attr_supported_features = cover.SUPPORT_OPEN | cover.SUPPORT_CLOSE | cover.SUPPORT_STOP
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, self._blind_group_obj.Hub.hub_ip)},
            name=self._blind_group_obj.Hub.name,
            manufacturer=MANUFACTURER,
            model=MODEL,
            via_device=(DOMAIN, self._blind_group_obj.Hub.hub_ip),
        )
```

iMicknl replied:
And perhaps you can even use the EntityDescription here, since most values are set to a static value anyways. But might be overkill if you only have a single cover configuration possible.

greg-ha-1990 replied:
Suggestion accepted, thanks.

iMicknl commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,180 @@
+            self._blind_group_id,
+        )
+
+    @property
+    def name(self):
+        """Name of the device."""
+        return self._blind_group_obj.name
+
+    @property
+    def unique_id(self):
+        """Return a unique ID for this device."""
+
+        return self._blind_group_id
```
```suggestion
```

greg-ha-1990 replied:
Suggestion accepted, thanks.

iMicknl commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,180 @@
+
+        return self._blind_group_id
+
+    @property
+    def assumed_state(self):
+        """Indicate that we do not go to the device to know its state."""
+
+        return False
+
```
```suggestion
```

False by default

greg-ha-1990 replied:
Suggestion accepted, thanks.

iMicknl commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,180 @@
+        """Indicate that we do not go to the device to know its state."""
+        return self._blind_group_obj.position
+
+    @property
+    def should_poll(self):
+        """Indicate that the device does not respond to polling."""
+
+        return True
+
```
```suggestion
```

True by default

iMicknl commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,180 @@
+
+        return True
+
+    @property
+    def supported_features(self):
+        """Bitmap indicating which features this device supports."""
+
+        return cover.SUPPORT_OPEN | cover.SUPPORT_CLOSE | cover.SUPPORT_STOP
+
+    @property
+    def device_class(self):
+        """Indicate we're managing a Roller blind motor group."""
+
+        return cover.DEVICE_CLASS_SHADE
+
+    @property
+    def device_info(self):
+        """Return the device_info of the device."""
+
+        device_info = {
+            "identifiers": {(DOMAIN, self._blind_group_obj.Hub.hub_ip)},
+            "name": self._blind_group_obj.Hub.name,
+            "manufacturer": MANUFACTURER,
+            "model": MODEL,
+            "via_device": (DOMAIN, self._blind_group_obj.Hub.hub_ip),
+        }
+
+        return device_info
```
```suggestion
```

greg-ha-1990 replied:
Suggestion accepted, thanks.

iMicknl reviewed (changes requested):
Thanks for your contribution! Initial review to get you started.

Have you tested this PR? Not sure if I overlook it, but doesn't seem you handle state updates / local polling anywhere?

greg-ha-1990 committed:
code optimization for leviosa shades

greg-ha-1990 commented:
> Thanks for your contribution! Initial review to get you started.
> 
> Have you tested this PR? Not sure if I overlook it, but doesn't seem you handle state updates / local polling anywhere?

This is a wifi to remote device, it can't actively know the state of the cover because there is no network connection between them, just as there is no way to ask the TV remote which channel is playing on the TV right now. We can only save the cached state values when actively controlling a wifi to IR device, and besides, cover can only be controlled by this IR device, so there's no ambiguity about it

greg-ha-1990 committed:
remove ambiguity for leviosa shades

iMicknl commented on homeassistant/components/leviosa_shades/manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "config_flow": true,
+  "dependencies": ["cover"],
+  "documentation": "https://www.home-assistant.io/integrations/leviosa_shades",
+  "iot_class": "local_push",
```
```suggestion
  "iot_class": "assumed_state",
```

This is why I was confused by no state in your integration. If you have a stateless integration (as you described in the reply, added below), your iot class is assumed state. See https://developers.home-assistant.io/docs/creating_integration_manifest/#iot-class.

> This is a wifi to remote device, it can't actively know the state of the cover because there is no network connection between them, just as there is no way to ask the TV remote which channel is playing on the TV right now. We can only save the cached state values when actively controlling a wifi to IR device, and besides, cover can only be controlled by this IR device, so there's no ambiguity about it.

greg-ha-1990 replied:
感谢您的建议，我之前忽略了assumed_state的选项，已更正，同时重新添加了以下属性：
```python
@property
def assumed_state(self):
    """Indicate that we do not go to the device to know its state."""
    return True
```

greg-ha-1990 replied:
Suggestion accepted, thanks.

greg-ha-1990 committed:
code optimization for leviosa shades

iMicknl commented on homeassistant/brands/leviosa.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "domain": "leviosa",
+  "name": "Leviosa",
+  "integrations": ["leviosa_shades"]
+}
```
Let's add this to a follow-up PR as well. This will also require a logo for `leviosa` domain. 

And since it seems Leviosa only offers one type of services, the Leviosa Shades, I am not sure if adding the brand adds any value here. See https://developers.home-assistant.io/docs/creating_integration_brand/.

greg-ha-1990 replied:
Suggestion accepted, thanks.

iMicknl commented on homeassistant/components/leviosa_shades/manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "name": "Leviosa Motor Shades",
+  "codeowners": ["@greg-ha-1990"],
+  "config_flow": true,
+  "dependencies": ["cover"],
```
```suggestion
```

greg-ha-1990 replied:
Suggestion accepted, thanks.

iMicknl commented on homeassistant/components/leviosa_shades/manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "codeowners": ["@greg-ha-1990"],
+  "config_flow": true,
+  "dependencies": ["cover"],
+  "documentation": "https://www.home-assistant.io/integrations/leviosa_shades",
```
```suggestion
  "documentation": "https://www.home-assistant.io/integrations/leviosa_shades",
   "integration_type": "hub",
```

See https://developers.home-assistant.io/docs/creating_integration_manifest/#integration-type. Have a look if hub or device is best in your situation.

greg-ha-1990 replied:
Suggestion accepted, thanks.

iMicknl commented on homeassistant/components/leviosa_shades/__init__.py:
```diff
@@ -0,0 +1,36 @@
+        hub_ip=hub_ip, hub_name=hub_name, websession=async_get_clientsession(hass)
+    )
+    await hub.getHubInfo()  # Check all is good
+    _LOGGER.debug("Hub object created, FW: %s", hub.fwVer)
```
Since you have details of your hub, you could add in a follow-up PR:

```python3
  device_registry.async_get_or_create(
            config_entry_id=entry.entry_id,
            identifiers={(DOMAIN, hub.mac)},
            manufacturer=MANUFACTURER,
            sw_version=hub.fwVer
        )
```

This will also nicely link the other entities to this hub, since you already use `via_device`. You can add a name/model to the hub as well, if you can detect this.

greg-ha-1990 replied:
OK I'll accepr your suggestion in next PR

iMicknl commented on homeassistant/components/leviosa_shades/__init__.py:
```diff
@@ -0,0 +1,36 @@
+    hub = tZoneHub(
+        hub_ip=hub_ip, hub_name=hub_name, websession=async_get_clientsession(hass)
+    )
+    await hub.getHubInfo()  # Check all is good
```
Here you would need to catch the exceptions as well and raise `ConfigEntryError` or `ConfigEntryNotReady`.

greg-ha-1990 replied:
Suggestion accepted, thanks.

iMicknl commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,141 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the Leviosa shade groups."""
+    _LOGGER.debug(
+        "Setting up %s[%s]: %s",
+        entry.domain,
+        entry.title,
+        entry.entry_id,
+    )
```
Do you really need this debug logging? By default there is already debug logging around setting up new platforms. I would try to keep debug logging to what is required to debug this integration.

greg-ha-1990 replied:
I will remove them

greg-ha-1990 replied:
Suggestion accepted, thanks.

iMicknl commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,141 @@
+        entry.entry_id,
+    )
+    hub: tZoneHub = hass.data[DOMAIN][entry.entry_id]
+    hub_mac = entry.data["device_mac"]
```
if you use "device_mac" multiple times, best is to use a constant.

greg-ha-1990 replied:
Suggestion accepted, thanks.

iMicknl reviewed (changes requested):
Thanks! Some additional comments.

greg-ha-1990 committed:
leviosa remove services

greg-ha-1990 committed:
reove brands file for leviosa shades

iMicknl commented on homeassistant/components/leviosa_shades/__init__.py:
```diff
@@ -0,0 +1,36 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Leviosa shades Zone from a config entry."""
+    hub_name = entry.title
+    hub_ip = entry.data["host"]
```
```suggestion
    hub_ip = entry.data[CONF_HOST]
```

greg-ha-1990 replied:
Suggestion accepted, thanks.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/config_flow.py:
```diff
@@ -0,0 +1,152 @@
+        return self.async_abort(reason="no_new_devs")
+
+    async def async_step_select(self, user_input=None):
+        """Handle multiple motion gateways found."""
```
The original design had it done for `one config entry to one gateway`. Because only one gateway can be bound at a time, the user can only click on one hub that he wants to bind at a time.

greg-ha-1990 replied:
I am not quite sure what you mean

iMicknl replied:
I misunderstood what happened in step_select, thought it was a multi select.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,136 @@
+        )
+
+    @property
+    def current_cover_position(self):
+        """Indicate that we do not go to the device to know its state."""
+        return self._blind_group_obj.position
```
Here is the property of the cover, not the state (the state is only on/off), however we do have no way of knowing the actual position of the cover, similar to the assumed_state, and for the position of the curtains we can only cache the value passed in during the last execution of the `move_next_position` or `move_next_position` method, similar to the state caches.

see:
![](https://jijiantuku-image.oss-cn-beijing.aliyuncs.com/markdown_img/2023/11/29/20231129_093915_a5278b81467b43fc41f8bbf54628a7f8.png)

greg-ha-1990 replied:
I have removed position releatedethods

greg-ha-1990 committed:
code optimization for leviosa shades

greg-ha-1990 commented on tests/components/leviosa_shades/__init__.py:
```diff
@@ -0,0 +1,13 @@
+
+def configure_integration(hass: HomeAssistant) -> MockConfigEntry:
+    """Configure the integration."""
+    entry = MockConfigEntry(domain=DOMAIN, entry_id="123456", unique_id="123456")
```
I just removed this test codes

greg-ha-1990 commented on tests/components/leviosa_shades/test_init.py:
```diff
@@ -0,0 +1,18 @@
+"""Tests for the SwitchBot Cloud integration init."""
```
Suggestion accepted, thanks.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,136 @@
+COVER_NEXT_POS_SCHEMA = vol.Schema({vol.Optional(ATTR_ENTITY_ID): cv.entity_ids})
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
```
Suggestion accepted, thanks.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,136 @@
+    """Represents a Leviosa shade group entity."""
+
+    def __init__(
+        self, hass: HomeAssistant, blind_group_id, blind_group_obj: tShadeGroup
```
Suggestion accepted, thanks.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,136 @@
+        return self._blind_group_obj.position
+
+    @property
+    def is_closed(self):
```
Suggestion accepted, thanks.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,136 @@
+
+class LeviosaBlindGroup(cover.CoverEntity):
+    """Represents a Leviosa shade group entity."""
+
```
Suggestion accepted, thanks.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,136 @@
+# Estimated time it takes to complete a transition
+# from one state to another
+TRANSITION_COMPLETE_DURATION = 30
+PARALLEL_UPDATES = 1
```
Suggestion accepted, thanks.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,136 @@
+        new_group_obj = hub.AddGroup(blind_group)
+        entities.append(
+            LeviosaBlindGroup(
+                hass, hub_mac + "-" + str(new_group_obj.number), new_group_obj
```
Suggestion accepted, thanks.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,136 @@
+        """Is the blind group currently closed?."""
+        return self._blind_group_obj.position == 0
+
+    async def async_close_cover(self, **kwargs):
```
Suggestion accepted, thanks.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,136 @@
+            | cover.CoverEntityFeature.STOP
+        )
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, self._blind_group_obj.Hub.hub_ip)},
```
My understanding is that this is only a LAN device and does not have a cloud service, so all communication depends on the correct IP. If the IP changes, it can only be re-binded. Unfortunately, leviosa does not provide a cloud service.
So we can only continue to use IP here.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,136 @@
+class LeviosaBlindGroup(cover.CoverEntity):
+    """Represents a Leviosa shade group entity."""
+
+    def __init__(
```
Suggestion accepted, thanks.

iMicknl commented on homeassistant/components/leviosa_shades/config_flow.py:
```diff
@@ -0,0 +1,152 @@
+            description_placeholders={"ip_add": self._host},
+        )
+
+    def _host_already_configured(self, host):
+        """See if we already have a hub with the host address configured."""
+        _LOGGER.debug("Checking if HOST was already configured")
+        existing_hosts = {
+            entry.data.get(CONF_HOST)
+            for entry in self._async_current_entries()
+            if CONF_HOST in entry.data
+        }
+        return host in existing_hosts
```
Can you use discovery to retrieve the mac address as well? I wouldn't check on host, but on mac address. If you set the mac address as unique id, you can use the following to check this:

```self._abort_if_unique_id_configured()```

iMicknl replied:
> Okay, so the question we're discussing is: what to do if the hub's IP changes?

We discuss two topics here:
- Use a real unique id (e.g. MAC) as unique ID
- How to handle IP changes

Will reply later in more detail, but point 1 is still valid. We don't see IP address as an unique property.

greg-ha-1990 replied:
> * Use a real unique id (e.g. MAC) as unique ID

We used `_host_id` as the `unique ID` before, as follows:
```python
await self.async_set_unique_id(self._host_uid)
```
Where the value of `_host_uid` is `uuid:f5825707-2a91-48d9-92ad-ec94cb68efa8` (as described in my reply above, which describes the details), and the **`ec94cb68efa8`** inside this string is the `MAC address`, preceded by some other information, so what you are saying is that we only need the `MAC address`, not the `uuid:f5825707-2a91-48d9-92ad-` that precedes it? But as far as I know there is no restriction on the setting of `unique_id`, as long as it uniquely represents the device.

Or maybe what you are trying to say is not really the `unqiue ID` setting, but the setting for `identifiers` in the following code:
```python
self._attr_device_info = DeviceInfo(
        identifiers={(DOMAIN, self._blind_group_obj.Hub.hub_ip)},
        name=self._blind_group_obj.Hub.name,
        manufacturer=MANUFACTURER,
        model=MODEL,
        via_device=(DOMAIN, self._blind_group_obj.Hub.hub_ip),
        )
```
The `identifiers` are set to IP here, do I need to change this to `MAC address`? if so can I do that too, just without calling `self._abort_if_unique_id_configured()` below.

greg-ha-1990 replied:
The identifiers have been changed to mac

greg-ha-1990 replied:
any oyher questions?

greg-ha-1990 replied:
@iMicknl This PR has been up for a long time again, what can I do to speed up the PR process?

davet2001 replied:
@greg-ha-1990 This PR has been open around ~1month which is not long in comparison with the list of >400 other open PRs.

I suggest reviewing other PRs is a good idea.

Also, with a big PR, a rebase then force push makes it harder to see the small edits after each review.

iMicknl reviewed (changes requested):
Thanks again! Some more comments.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/__init__.py:
```diff
@@ -0,0 +1,40 @@
+            hub_ip=hub_ip, hub_name=hub_name, websession=async_get_clientsession(hass)
+        )
+        await hub.getHubInfo()  # Check all is good
+    except Exception as err:
+        raise ConfigEntryError("get hub info failed") from err
```
I know that `catch Exception` is a very ugly way of writing.
 In fact, the initial version does not need to try to catch anything, because the constructor of `LeviosaZoneHub` and the method `getHubInfo` should handle all exceptions themselves.
My understanding is that `async_setup_entry` in `__init__.py` has already obtained the IP of the hub. If communication with this IP fails thereafter, then we have no way to determine whether the IP of the hub has changed or the hub is not powered on. Therefore, there is no way to distinguish between `ConfigEntryNotReady` and `ConfigEntryAuthFailed`.

greg-ha-1990 replied:
Suggestion accepted, thanks.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/config_flow.py:
```diff
@@ -0,0 +1,152 @@
+        """Allow user to enter details for a Leviosa Zone."""
+        errors = {}
+        if user_input is not None:
+            _LOGGER.debug(
+                "Config User step - validate and save [%s] @%s",
+                self._host_uid,
+                self._host,
+            )
+            for i in user_input:
+                _LOGGER.debug("UI %s -> %s", i, user_input[i])
```
not necessary, I'll remove them.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/config_flow.py:
```diff
@@ -0,0 +1,152 @@
+            elif fw_ver is None:
+                errors["base"] = "unknown"
+            else:
+                _LOGGER.debug("Saving Integration data")
```
not necessary, I'll remove them.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,108 @@
+
+        self._attr_name = self._blind_group_obj.name
+        self._attr_unique_id = self._blind_group_id
+        self._attr_device_class = cover.CoverDeviceClass.SHADE
+        self._attr_supported_features = (
+            cover.CoverEntityFeature.OPEN
+            | cover.CoverEntityFeature.CLOSE
+            | cover.CoverEntityFeature.STOP
+        )
```
Suggestion accepted, thanks.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,108 @@
+        """Indicate that we do not go to the device to know its state."""
+        return self._blind_group_obj.position
+
+    @property
+    def is_closed(self) -> bool:
+        """Is the blind group currently closed?."""
+        return self._blind_group_obj.position == 0
```
OK, I'll remove thoes properties.
They do tend to cause ambiguity

greg-ha-1990 replied:
I can not remove `is_closed`.
once I removed them, I got those errors:
```
Error adding entities for domain cover with platform leviosa_shades
Traceback (most recent call last):
  File "/data/ha/hacore/homeassistant/helpers/entity_platform.py", line 507, in async_add_entities
    await asyncio.gather(*tasks)
  File "/data/ha/hacore/homeassistant/helpers/entity_platform.py", line 752, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/data/ha/hacore/homeassistant/helpers/entity.py", line 1023, in add_to_platform_finish
    self.async_write_ha_state()
  File "/data/ha/hacore/homeassistant/helpers/entity.py", line 743, in async_write_ha_state
    self._async_write_ha_state()
  File "/data/ha/hacore/homeassistant/helpers/entity.py", line 843, in _async_write_ha_state
    state, attr = self._async_generate_attributes()
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/ha/hacore/homeassistant/helpers/entity.py", line 784, in _async_generate_attributes
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/ha/hacore/homeassistant/helpers/entity.py", line 749, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/data/ha/hacore/homeassistant/components/cover/__init__.py", line 273, in state
    if (closed := self.is_closed) is None:
                  ^^^^^^^^^^^^^^
  File "/data/ha/hacore/homeassistant/components/cover/__init__.py", line 328, in is_closed
    return self._attr_is_closed
           ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'LeviosaBlindGroup' object has no attribute '_attr_is_closed'
2023-11-29 20:54:11.077 ERROR (MainThread) [homeassistant.components.cover] Error while setting up leviosa_shades platform for cover
```

greg-ha-1990 replied:
any oyher questions?

greg-ha-1990 commented on homeassistant/components/leviosa_shades/__init__.py:
```diff
@@ -0,0 +1,40 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
```
Suggestion accepted, thanks.

greg-ha-1990 commented on homeassistant/components/leviosa_shades/__init__.py:
```diff
@@ -0,0 +1,40 @@
+    hub_name = entry.title
+    hub_ip = entry.data[CONF_HOST]
+    try:
+        hub = tZoneHub(
+            hub_ip=hub_ip, hub_name=hub_name, websession=async_get_clientsession(hass)
+        )
```
Suggestion accepted, thanks.

greg-ha-1990 committed:
code optimization for leviosa shades

greg-ha-1990 committed:
leviosa shades code optimization

greg-ha-1990 commented on homeassistant/components/leviosa_shades/config_flow.py:
```diff
@@ -0,0 +1,152 @@
+                errors["base"] = "unknown"
+            else:
+                _LOGGER.debug("Saving Integration data")
+                await self.async_set_unique_id(self._host_uid)
```
any oyher questions?

emontnemery commented on homeassistant/components/leviosa_shades/const.py:
```diff
@@ -0,0 +1,35 @@
+
+MANUFACTURER: Final = "Leviosa Motor Shades LLC"
+MODEL: Final = "Zone Hub"
+DEVICE_NAME: Final = "device_name"
```
This doesn't seem to be used, please remove it

dpiessens commented on homeassistant/components/leviosa_shades/config_flow.py:
```diff
@@ -0,0 +1,144 @@
+            self._host_uid = keys[idx_of_ip]
+            return await self.async_step_connect()
+
+        select_schema = vol.Schema(
+            {vol.Required("select_ip"): vol.In(self._devices.values())}
+        )
```
The hub itself doesn't have any S/N or MAC printed on it. Just a general brand label and required FCC markings.

dpiessens commented:
Hi, I now have 15 Leviosa shades in my house and effectively loaded this PR into my instance as this code works much better than what Leviosa has posted on their site. What's needed to get this pushed over the finish line? @greg-ha-1990 do you need help with any final changes / edits? @emontnemery I appreciate you wanting to understand how this is working, it is an odd setup. Any compromises we can make so this can get into the official release?

greg-ha-1990 commented on homeassistant/components/leviosa_shades/cover.py:
```diff
@@ -0,0 +1,90 @@
+class LeviosaBlindGroup(cover.CoverEntity):
+    """Represents a Leviosa shade group entity."""
+
+    _attr_device_class = cover.CoverDeviceClass.SHADE
```
Why should we remove _attr_device_class = cover.CoverDeviceClass.SHADE?

emontnemery replied:
That was a mistake, I meant to add the assume_state

greg-ha-1990 commented on homeassistant/components/leviosa_shades/const.py:
```diff
@@ -0,0 +1,35 @@
+DEVICE_FW_V: Final = "firmware"
+DEVICE_MAC: Final = "device_mac"
+
+BLIND_GROUPS: Final = "blind_groups"
+GROUP1_NAME: Final = "grp1_name"
+GROUP2_NAME: Final = "grp2_name"
+GROUP3_NAME: Final = "grp3_name"
+GROUP4_NAME: Final = "grp4_name"
+GROUP5_NAME: Final = "grp5_name"
+GROUP6_NAME: Final = "grp6_name"
```
I don’t think this is a good idea. My principle is that all const constants should be written in const.py. This is my programming habit. These constants will be used in other files in subsequent PRs that support service. . If the official flavor of HA in the future is "Only constants imported by two or more files can be written to const.py", then I will comply. But now, I still have to write all final constants into const,py

emontnemery replied:
>  If the official flavor of HA in the future is "Only constants imported by two or more files can be written to const.py"

That's the rule, yes

greg-ha-1990 commented on homeassistant/components/leviosa_shades/const.py:
```diff
@@ -0,0 +1,35 @@
+
+DOMAIN: Final = "leviosa_shades"
+
+MANUFACTURER: Final = "Leviosa Motor Shades LLC"
+MODEL: Final = "Zone Hub"
```
Because it is only referenced by cover.py, it must be written in cover.py? I do not understand why? I will add service to my next PR, and I will still use the constants in const.py at that time.

MaximUltimatum commented:
As someone who purchased leviosa shades and the "leviosa zone", I appreciate work being done to make sure this new integration meets home assistant standards. I would rather wait longer for my shades to work with home assistant than add extra technical debt into the home assistant maintainers

> The hub itself doesn't have any S/N or MAC printed on it. Just a general brand label and required FCC markings.

The "leviosa zone" is, as far as I can tell from examining mine, a repackaged IR blaster. It has a MAC address, and would require minimal effort on the part of the Leviosa team to add a sticker to it with the devices Mac address (or, worst case, users can just find it on their router as I have)
It seems silly (to me) to add a bunch of extra user pain trying to track what DHCP has assigned a device's IP to... when a simple process improvement by switching to using a universal identifier solves this problem as it's designed to

emontnemery commented:
@dpiessens there are many unaddressed review comments. If @greg-ha-1990 doesn't have the time to address them, someone else needs to step up for this PR to be merged. An alternative is to publish a custom integration.

greg-ha-1990 commented:
> @dpiessens there are many unaddressed review comments. If @greg-ha-1990 doesn't have the time to address them, someone else needs to step up for this PR to be merged. An alternative is to publish a custom integration.

I didn't see this rule in the official recommended style. 

In fact, if you do a global search for the key word `: Final = ` in the core repository, you'll see that most of the results are within `const.py`.

![](https://jijiantuku-image.oss-cn-beijing.aliyuncs.com/markdown_img/2024/04/17/20240417_085218_c93c9b93158827b1a5bd4fc51ca268f3.png)

I just looked for a few other integrations as a reference. I can easily find that their constants in `const.py` have only been imported once, such as:
`ATTRIBUTION` in [adobe](https://github.com/home-assistant/core/blob/dev/homeassistant/components/abode/const.py), which is only used in `__init__.py`, Another example is `ATTRIBUTION` in [aftership](https://github.com/home-assistant/core/blob/dev/homeassistant/components/aftership/const.py), which is only used by `sensor.py` . You can find tons of examples of this if you want.

It should be written in const.py as long as it has the `public` constant attribute, and if it has to be set to private after being imported only once, it will inevitably make maintenance difficult, because the entity code is iterated and may be maintained by different people, and in that case, it will inevitably need a public const.py to life the public constants, which is what most integrations do.

emontnemery commented:
@greg-ha-1990 apart from the style issue about non shared constants in `const.py`, there are many other comments which you have not addressed.
